templates/model/validate.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	"golang.org/x/exp/slices"
+	"gopkg.in/yaml.v3"
+)
+
+func notZero[T comparable](pos ConfigPos, x valWithPos[T], fieldName string) error {
+	var zero T
+	if x.Val == zero {
+		return pos.AnnotateErr(fmt.Errorf("field %q is required", fieldName))
+	}
+	return nil
+}
+
+func nonemptySlice[T any](pos ConfigPos, s []T, fieldName string) error {
+	if len(s) == 0 {
+		return pos.AnnotateErr(fmt.Errorf("field %q is required", fieldName))
+	}
+	return nil
+}
+
+// Fail if any unexpected fields are seen. The input must be a mapping/object; anything else will
+// result in an error.
+// This is a workaround for the brokenness of KnownFields in the upstream yaml lib
+// (https://github.com/go-yaml/yaml/issues/460).
+func extraFields(n *yaml.Node, knownFields []string) error {
+	if n.Kind != yaml.MappingNode {
+		return fmt.Errorf("got yaml node of kind %d, expected %d", n.Kind, yaml.MappingNode)
+	}
+	m := map[string]any{}
```

Comment:
I don't know if we can guarantee that the response is a `map[string]any`? It could be a `[]any` too.

Reply:
The check for MappingNode three lines up actually guarantees this.